JitifyException in cupyx.scipy.ndimage.gaussian_filter with higher dimensions//large arrays?

### Description



```
_=cupyx.scipy.ndimage.gaussian_filter(cp.random.randn(26,3,52,70,50,dtype='float32'), [1,1,1,1,1], mode='nearest')

```
works fine, but 
```
_=cupyx.scipy.ndimage.gaussian_filter(cp.random.randn(26,3,52,70,50,70,dtype='float32'), [1,1,1,1,1,1], mode='nearest')

```
results  in a  

```
JitifyException                           Traceback (most recent call last)
File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/IPython/core/interactiveshell.py, in run_code:
Line 3441:  exec(code_obj, self.user_global_ns, self.user_ns)

In  [30]:
Line 1:     _=cupyx.scipy.ndimage.gaussian_filter(cp.random.randn(26,3,52,70,50,70,dtype='float32'), [1.,2.,1.,1.,2.,1.], mode='nearest')

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/filters.py, in gaussian_filter:
Line 367:   mode, cval, 0)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/filters.py, in _run_1d_correlates:
Line 207:   input, wghts, output, mode, cval, origin)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/_filters_core.py, in _run_1d_filters:
Line 97:    fltr(input, arg, axis, output, mode, cval, origin)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/filters.py, in correlate1d:
Line 113:   return _correlate_or_convolve(input, weights, output, mode, cval, origins)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/filters.py, in _correlate_or_convolve:
Line 177:   weights_dtype=weights_dtype)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupyx/scipy/ndimage/_filters_core.py, in _call_kernel:
Line 141:   kernel(*args)

File cupy/_core/_kernel.pyx, in cupy._core._kernel.ElementwiseKernel.__call__:
Line 850:   

File cupy/_core/_kernel.pyx, in cupy._core._kernel.ElementwiseKernel._get_elementwise_kernel:
Line 875:   

File cupy/_util.pyx, in cupy._util.memoize.decorator.ret:
Line 59:    

File cupy/_core/_kernel.pyx, in cupy._core._kernel._get_elementwise_kernel:
Line 662:   

File cupy/_core/_kernel.pyx, in cupy._core._kernel._get_simple_elementwise_kernel:
Line 62:    

File cupy/_core/core.pyx, in cupy._core.core.compile_with_cache:
Line 2042:  

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupy/cuda/compiler.py, in compile_with_cache:
Line 463:   cache_in_memory, jitify)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupy/cuda/compiler.py, in _compile_with_cache_cuda:
Line 540:   log_stream, cache_in_memory, jitify)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupy/cuda/compiler.py, in compile_using_nvrtc:
Line 299:   name_expressions, log_stream, jitify)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupy/cuda/compiler.py, in _compile:
Line 268:   source, options, cu_path)

File /fast/work/users/hoffmmax_c/anaconda3/envs/img/lib/python3.7/site-packages/cupy/cuda/compiler.py, in _jitify_prep:
Line 254:   raise JitifyException(str(cex))

JitifyException: Runtime compilation failed
```

Version:



### To Reproduce

```py

import cupy as cp
import cupyx
_=cupyx.scipy.ndimage.gaussian_filter(cp.random.randn(26,3,52,70,50,dtype='float32'), [1,1,1,1,1], mode='nearest')
```


### Installation

Wheel (`pip install cupy-***`)

### Environment

```
OS                           : Linux-3.10.0-1160.41.1.el7.x86_64-x86_64-with-centos-7.9.2009-Core
Python Version               : 3.7.9
CuPy Version                 : 9.5.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.19.4
SciPy Version                : 1.6.0
Cython Build Version         : 0.29.22
Cython Runtime Version       : None
CUDA Root                    : /usr/local/cuda-10.2
nvcc PATH                    : /usr/local/cuda-10.2/bin/nvcc
CUDA Build Version           : 10020
CUDA Driver Version          : 11040
CUDA Runtime Version         : 10020
cuBLAS Version               : (available)
cuFFT Version                : 10501
cuRAND Version               : 10205
cuSOLVER Version             : (10, 3, 0)
cuSPARSE Version             : (available)
NVRTC Version                : (10, 2)
Thrust Version               : 100907
CUB Build Version            : 100800
Jitify Build Version         : 60e9e72
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla V100-SXM2-32GB
Device 0 Compute Capability  : 70
Device 0 PCI Bus ID          : 0000:18:00.0
```

### Additional Information

```
_=cupyx.scipy.ndimage.gaussian_filter(cp.random.randn(26,3,52,70,50,20,dtype='float32'), [1,1,1,1,1,1], mode='nearest')
```

works, too. So it is just related to the size of the array in relation to GPU memory? In this case, it might be cool to have a more informative stack trace

> To me there shouldn't be any ambiguity (`(long, long)` is the only allowed overload)...

I see. Even for host compiler it's ambiguous:
```
$ g++ --std=c++11 test_std_max.cpp -o test_std_max
test_std_max.cpp: In function ‘int main()’:
test_std_max.cpp:8:21: error: no matching function for call to ‘max(ptrdiff_t&, int&)’
    8 |     cout << max(a, b) << endl;
      |                     ^
In file included from /usr/include/c++/9/bits/char_traits.h:39,
                 from /usr/include/c++/9/ios:40,
                 from /usr/include/c++/9/ostream:38,
                 from /usr/include/c++/9/iostream:39,
                 from test_std_max.cpp:1:
/usr/include/c++/9/bits/stl_algobase.h:222:5: note: candidate: ‘template<class _Tp> const _Tp& std::max(const _Tp&, const _Tp&)’
  222 |     max(const _Tp& __a, const _Tp& __b)
      |     ^~~
/usr/include/c++/9/bits/stl_algobase.h:222:5: note:   template argument deduction/substitution failed:
test_std_max.cpp:8:21: note:   deduced conflicting types for parameter ‘const _Tp’ (‘long int’ and ‘int’)
    8 |     cout << max(a, b) << endl;
      |                     ^
In file included from /usr/include/c++/9/bits/char_traits.h:39,
                 from /usr/include/c++/9/ios:40,
                 from /usr/include/c++/9/ostream:38,
                 from /usr/include/c++/9/iostream:39,
                 from test_std_max.cpp:1:
/usr/include/c++/9/bits/stl_algobase.h:268:5: note: candidate: ‘template<class _Tp, class _Compare> const _Tp& std::max(const _Tp&, const _Tp&, _Compare)’
  268 |     max(const _Tp& __a, const _Tp& __b, _Compare __comp)
      |     ^~~
/usr/include/c++/9/bits/stl_algobase.h:268:5: note:   template argument deduction/substitution failed:
test_std_max.cpp:8:21: note:   deduced conflicting types for parameter ‘const _Tp’ (‘long int’ and ‘int’)
    8 |     cout << max(a, b) << endl;
      |                     ^
```

--- a/cupyx/scipy/ndimage/_filters_core.py
+++ b/cupyx/scipy/ndimage/_filters_core.py
@@ -225,7 +225,7 @@ def _generate_nd_kernel(name, pre, found, post, mode, w_shape, int_type,
                          format(j=j, type=int_type))
         else:
             boundary = _util._generate_boundary_condition_ops(
-                mode, 'ix_{}'.format(j), 'xsize_{}'.format(j))
+                mode, 'ix_{}'.format(j), 'xsize_{}'.format(j), int_type)
             # CArray: last line of string becomes inds[{j}] = ix_{j};
             loops.append('''
     for (int iw_{j} = 0; iw_{j} < {wsize}; iw_{j}++)

--- a/cupyx/scipy/ndimage/_util.py
+++ b/cupyx/scipy/ndimage/_util.py
@@ -117,8 +117,11 @@ def _generate_boundary_condition_ops(mode, ix, xsize, int_t="int",
         }}'''.format(ix=ix, xsize=xsize, min=min_func)
     elif mode == 'nearest':
         ops = '''
-        {ix} = {min}({max}({ix}, 0), {xsize} - 1);'''.format(
-            ix=ix, xsize=xsize, min=min_func, max=max_func)
+        {ix} = {min}({max}(({T}){ix}, ({T})0), ({T})({xsize} - 1));'''.format(
+            ix=ix, xsize=xsize, min=min_func, max=max_func,
+            # force using 64-bit signed integer for ptrdiff_t,
+            # see cupy/cupy#6048
+            T=('int' if int_t == 'int' else 'long long'))
     elif mode == 'grid-wrap':
         ops = '''
         {ix} %= {xsize};

--- a/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
+++ b/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
@@ -228,6 +228,17 @@ class TestFilter(FilterTestCaseBase):
         return self._filter(xp, scp)
 
 
+@testing.with_requires('scipy')
+class TestNearestFilterEdgeCase:
+
+    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
+    def test_filter(self, xp, scp):
+        # Use a large input to check any compilation error, see cupy/cupy#6048
+        return scp.ndimage.gaussian_filter(
+            testing.shaped_random((26, 3, 52, 70, 50), xp, xp.float32),
+            [1, 1, 1, 1, 1], mode='nearest')
+
+
 def dummy_deriv_func(input, axis, output, mode, cval, *args, **kwargs):
     # For testing generic_laplace and generic_gradient_magnitude. Doesn't test
     # mode, cval, or extra argument but those are tested indirectly with

